src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -253,6 +253,30 @@ public boolean isNative() {
         return false;
     }
 
+    @Override
+    public boolean isOverlapping(MemorySegment other) {
+        AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)Objects.requireNonNull(other);
+        if (base() == that.base()) {  // both are either native or heap
+            this.checkAccess(0, this.length, true);
```

Comment:
I don't think `checkAccess` is needed? That operation is typically called when accessing a segment to dereference, to verify that:
* the accessed location is within the bounds of the segment
* the access mode (read/write) is compatible with the segment
Here, you don't do dereference, so these checks are not necessary, `checkValidState()` is what you want.